lib/nmatrix/shortcuts.rb

```diff
@@ -322,6 +322,76 @@ def random(shape, opts={})
 
     #
     # call-seq:
+    #     random_ortho(shape) -> NMatrix
+    #
+    # Randomly generates a +:dense+ orthogonal NMatrix, or a unitary complex NMatrix if dtype is complex. 
+    # The parameter is the dimension of the matrix. For any non-integer dtype the algorithm creates a Householder 
+    # matrix from random numbers between 0 and 1 generated by +Random::rand+. For any integer dtype an orthogonal
+    # matrix is a random column permutation of a diagonal matrix with entries equal to +1 or -1 randomly 
+    # distributed along its diagonal. You may not currently generate random numbers for a rational matrix.
+    #
+    # * *Arguments* :
+    #   - +shape+ -> Array or integer specifying the dimensions 
+    #   (note: an othogonal matrix must be square by definition).
+    # * *Returns* :
+    #   - A randomly generated orthogonal NMatrix (unitary complex for complex dtype).
+    #
+    # Examples:
+    #
+    #  m = NMatrix.rand_ortho(3) #  => -0.81460661007588 -0.05620899291406391 -0.5772838296161336
+    #                                 -0.05620899291406391 0.9982588783338108 -0.017881860733958227
+    #                                 -0.5772838296161336 -0.017881860733958227 0.8163477317420696 
+    #  m.dot m.transpose #  => 0.9999999999999991 -2.6020852139652106e-17 -2.7755575615628914e-16
+    #                          -2.6020852139652106e-17 1.0                -3.469446951953614e-18
+    #                          -2.7755575615628914e-16 -3.469446951953614e-18 1.0
+    #  m.transpose.dot m #  => 0.9999999999999991 -2.6020852139652106e-17 -2.7755575615628914e-16
+    #                         -2.6020852139652106e-17 1.0                 -3.469446951953614e-18
+    #                         -2.7755575615628914e-16 -3.469446951953614e-18 1.0
+    # References:
+    # 
+    #   Stewart 1980, "The Efficient Generation of Random Orthogonal Matrices 
+    #   with an Application to Condition Estimators"
+    #
+    def random_ortho(shape, opts={})
+      raise(NotImplementedError, "does not support rational random number generation") if [:rational64, :rational128].include? opts[:dtype]
+
+      if shape.is_a?(Fixnum)
+        n = shape 
+      elsif shape.is_a?(Array)
+        raise(ShapeError, "Expected 2D matrix") unless shape.length == 2
+        raise(ShapeError, "Expected square matrix") unless (shape[0] == shape[1])
+        n = shape[0] 
+      else
+        raise(ShapeError, "Expected 2D matrix")
+      end
+
+      # integer valued orthogonal
+      if [:byte, :int8, :int16, :int32, :int64].include? opts[:dtype]
+        d = n.times.map { [-1,1].sample }
+        q = NMatrix.diagonal(d)
+        rand_order = (0...n).to_a.sample(n)
+        q.permute_columns!(rand_order)
```

Comment:
@cjfuller I assumed that column permutations work in NMatrix the same way they work in Matlab or R (otherwise NMatrix might be unnecessarily confusing for scientist having experience with Matlab or R). Is that not true? 

That is, the required permutation for your example would be `p=[2,1,0]`. 
Equivalent in Matlab: `m=diag(1:3); m(:, [3,2,1])`
Equivalent in R: `m=diag(1:3), m[, c(3,2,1)]`
And at least for the given example it also works in NMatrix the same way:
`[17] pry(main)> m = N.diagonal([1,2,3])`
`=> [  [1, 0, 0]   [0, 2, 0]   [0, 0, 3] ]`
`[18] pry(main)> p = [2,1,0]`
`=> [2, 1, 0]`
`[19] pry(main)> m.permute_columns(p)`
`=> [  [0, 0, 1]   [0, 2, 0]   [3, 0, 0] ]`